cfgrammar/src/lib/yacc/ast.rs

```diff
@@ -6,16 +6,109 @@ use std::{
+    yacckind: YaccKind,
+    ast: GrammarAST,
+    any_errors_found: bool,
+}
+
 /// Contains a `GrammarAST` structure produced from a grammar source file.
 /// As well as any errors which occurred during the construction of the AST.
 pub struct ASTWithValidityInfo {
     ast: GrammarAST,
-    errs: Vec<YaccGrammarError>,
+    report: DedupReport<SimpleReport>,
+}
+
+impl<'a, R> ASTBuilder<'a, R>
+where
+    R: ErrorReport,
+{
+    /// Returns a new `ASTBuilder` with `source`, and an optional `path`
+    /// The path may be used for error reporting, but is not otherwise used for any
+    /// filesystem operations or checked to be a valid path.
+    pub fn with_source(source: &'a str) -> Self {
+        Self {
+            source,
+            yacckind: None,
+            report_handler: None,
+        }
+    }
+    pub fn yacckind(mut self, yk: YaccKind) -> Self {
+        self.yacckind = Some(yk);
+        self
+    }
+    pub fn error_report(mut self, report: &'a mut R) -> Self {
+        self.report_handler = Some(ReportHandler::new(report));
+        self
+    }
+
+    /// Builds an `ast`, and returns a `GrammarBuilder`
+    pub fn grammar_builder(mut self) -> Result<GrammarBuilder, ASTBuilderError> {
+        let report_handler = self
+            .report_handler
+            .as_mut()
+            .ok_or(ASTBuilderError::MissingErrorReport)?;
+        let yacckind = self.yacckind.ok_or(ASTBuilderError::MissingYaccKind)?;
+
+        let ast = match yacckind {
+            YaccKind::Original(_) | YaccKind::Grmtools | YaccKind::Eco => {
+                let mut yp = YaccParser::new(yacckind, self.source.to_string());
+                yp.parse_with_report(report_handler).ok();
+                let mut ast = yp.ast();
+                ast.complete_and_validate()
+                    .map_err(|e| report_handler.non_fatal_error(e))
+                    .ok();
+                ast
+            }
+        };
+        ast.unused_symbols().for_each(|symidx| {
+            let (kind, span) = match symidx.symbol(&ast) {
+                Symbol::Rule(_, span) => (YaccGrammarWarningKind::UnusedRule, span),
+                Symbol::Token(_, span) => (YaccGrammarWarningKind::UnusedToken, span),
+            };
+            report_handler.warning(YaccGrammarWarning {
+                kind,
+                spans: vec![span],
+            })
+        });
+        report_handler.finish();
+        let any_errors_found = report_handler.any_errors_found();
+        Ok(GrammarBuilder {
+            yacckind,
+            ast,
+            any_errors_found,
+        })
+    }
+}
+
+impl GrammarBuilder {
+    pub fn build<StorageT>(&self) -> Result<YaccGrammar<StorageT>, ASTBuilderError>
+    where
+        StorageT: num_traits::PrimInt + num_traits::Unsigned + 'static,
+        usize: num_traits::AsPrimitive<StorageT>,
+    {
+        YaccGrammar::new_from_grammar_builder(self.yacckind, self)
+    }
+    pub fn any_errors_found(&self) -> bool {
```

Comment:
Dumb question: what is the scope of the "any errors"? With a builder, I think a normal Rust expectation is that nothing much happens until `build` is called, at which point `self` tends to be consumed and one gets out another thing entirely. Here we have `build(&self)` so I'm not quite sure what the expected chain is.

Reply:
It is basically the negation of `ASTWithValidationInfo::is_valid`, so the scope is during `GrammarAST` construction,
I think you had used the term in one of the previous PRs where we were discussing `ASTBuilder` so we could keep it the same across those two if you prefer?

Reply:
But yeah, I believe the reasons it is taking `&self` were historical, and no longer relevant.
It seems fine to take `self` here to me.

Reply:
Actually nevermind, the real reason it doesn't take self is because the user may want to call `grammar_builder.ast()` in response to an error e.g. to format conflicts.  But there is nothing here exercising that yet.

Reply:
I'm inclined to think that we should bite the bullet on this sort of thing and work out what the "best" API would be and worry about backwards compatibility afterwards. I suspect that in most cases the breakage will probably be minimal, and perhaps non-existent.

Reply:
Interestingly `any_errors_found()` isn't needed for backwards compatibility, where it uses `self.errors().is_empty()`.
It is the movement of ownership of `errors` field into an `ErrorReport` that causes us to need that boolean (for `YaccGrammar::new_from_grammar_builder`.

My lsp is very likely the only thing actually using these APIs though. I think we could make `pub(crate) any_errors_found()` instead of `pub`.

But having the `ast` even if we can avoid it entirely for error formatting of things like conflicts in the future, still seems pretty crucial for many lsp needs in particular this would mostly surface in things like *jump to rule*.  Anyhow i'd be happy to remove e.g. `ASTWithValidationInfo` and `YaccGrammar::new_from_ast_with_validation_info` as I think I'm the only one currently using *both* the *ast* and the *grammar* outside the libraries.

Reply:
So, if we made `any_errors_found` private, that just leaves us with `ast()` and `build()`, we could rename `ast()` `ast_ref()` perhaps, and replace `build()` with `TryFrom`.  Which to me is starting to sound something like an `GrammarAstRef`.  I think the key weirdness here is that we have an *Owned* (currently) `GrammarBuilder` value, which only gives references to it's inner value `&GrammarAST`.  Otherwise `GrammarAST` could be mutated and we would no longer know if it is valid.

I think these are some of the key points about why this API seems a little awkward?

Reply:
> I think these are some of the key points about why this API seems a little awkward?

Yes, I think so.

Can we reimagine the API at an even higher level perhaps? [I don't have immediate ideas on this I must admit!]

Reply:
Or maybe even at a lower level, like we could have a trait or another type that gives an owned variation of `GrammarAST` but without any `pub` fields or `&mut` functions, then it would be pretty natural to use/pass references to it?

If we did it as a trait we wouldn't necessarily have to add a new type. For example we could return something like `impl ReadonlyGrammarAST`. That leaves open the question of validity, would it make sense to have a `is_valid()` function which would only allow us to return them after `complete_and_validate`?

Reply:
The other thing that strikes me is that perhaps just a terminology change might help, such as we could rename `grammar_builder()` to `validate()`, which could return perhaps a `ASTReceipt` instead of the nested `GrammarBuilder`.
from there we could have `TryFrom` or `receipt.grammar()`

Reply:
> The other thing that strikes me is that perhaps just a terminology change might help, such as we could rename grammar_builder() to validate()

Yes, I think that's probably a good idea! That could easily a PR in its own right I guess?

Reply:
It kind of depends I think, if we do that first and then decide whether we want to change the ownership situation of errors later?  Then it is essentially a renaming and just changing where one function is `pub`.

If we view `ASTBuilder` (e.g. as in this PR) a rename of `ASTWithValidationInfo`, I think we have a few things:
* Rename, [easy]
* Move where `YaccGrammar::new_from_ast_with_validation_info` is pub into the chain. [easy]
* Deprecation and removal of `errors()` and `warnings()` which is the primary difficulty that causes all the 7 or so patches of new code before I introduce it in this series.

Reply:
So, I thought about your notion of needing to perhaps think about how we can fix things at a higher level.
What I think a higher level notion we might be missing here is a `Driver`.

I also am kind of doubting the entire builder-esque pattern is a good fit for this instance, if we look at the fields in my `ASTBuilder` impl they are actually all required, and we're returning errors when any are none, which seems weird.
I.e. it's just introducing potential errors, so we can use method chaining basically.  

I was looking through the [init-struct-pattern](https://xaeroxe.github.io/init-struct-pattern/) there is another pattern not shown in that link, but is similar since it has an init function, and `Default` implementing field. That allows for required options, and seems nice to me.

I took a stab at a mockup implementation of how a notion of a `Driver` could work using the `Report` trait idea.
I think part of the problem with this builder notion is also that results break the chain, but we have multiple values we may want to return.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=df2df54c4578dda35d7bc5d2b6698bdc

But with this driver notion, I think the let binding ends up working pretty well, since we've taken the field construction out of the method chain.  Curious if you have any thoughts?

Reply:
FWIW, I did do some experimenting here with the notion of a cross-tool driver:

Which tries to abstract the notion of lrpar/lrlex into a `Tool`, and define a abstraction boundary to run them,
managing source file loading, and error reporting (There are a bunch of renames there `Report` changed to `Diagnostics` and `ReportHander` to `DiagnosticsEmitter`,  then `finished()` to `no_more_data()`).

https://github.com/ratmice/driver/

When I try and ask myself "What problem is this trying to solve", the answer that I have is that it is trying to bridge the `CTParserBuilder` and `CTLexerBuilder` so that the stuff which is currently locked away behind within the `CTParserBuilder::build()` closure, can be used by `CTLexerBuilder`, e.g. in the `missing_from_lexer()` check which cannot happen until *after* the parser is built, we have access to the data about the symbol in the parser that is missing.

Overall the `Driver` implementation turned out a bit more complex than I had hoped, mostly so we aren't baking in choices that we don't have a mechanism to expand upon.  I'm not certain I was entirely successful at baking those choices in.  But overall, it *is* a different way to look at things than this entire series.

In general if we were to go this route `CTLexerBuilder` would own a parser driver, and be able to extract the `options` from the `CTParserBuilder` returned from `lrpar_config`.  Which it could then send to the `Driver` instead of calling `build`.

I really don't know if this is going to be fruitful, but I do think it highlights part of the issue here being abstraction boundaries, and passing mutable state as a way to avoid giving up ownership across those abstraction boundaries.